platform/mbed_wait_api_rtos.cpp

```diff
@@ -31,23 +31,39 @@ void wait(float s)
 
 void wait_ms(int ms)
 {
-    wait_us(ms * 1000);
+    // Wait equivalent to 1 ms near the tick might result into zero wait.
+    // Hence ms precision can be used for > 1 ms wait only.
+    if ((ms >= 1) && core_util_are_interrupts_enabled()) {
+        ThisThread::sleep_for(ms);
+    } else {
+        wait_us(ms * 1000);
+    }
 }
 
 void wait_us(int us)
 {
-    const ticker_data_t *const ticker = get_us_ticker_data();
-
-    uint32_t start = ticker_read(ticker);
-    if ((us >= 1000) && core_util_are_interrupts_enabled()) {
+    uint32_t ms = 0;
+    uint32_t sub_ms = us;
+    // Wait > 10ms will have ms precision, else us
+    if ((us >= 10000) && core_util_are_interrupts_enabled()) {
         // Use the RTOS to wait for millisecond delays if possible
-        sleep_manager_lock_deep_sleep();
-        Thread::wait((uint32_t)us / 1000);
-        sleep_manager_unlock_deep_sleep();
+        ms = (us / 1000);
```

Comment:
you could do the rounding as part of this statement:
```c
ms = ((us + 500) / 1000);
```

Reply:
I'm not really seeing the motivation for changing `wait_us`. If someone is calling it specifically, they want the precision, or else they'd be calling `wait_ms`, right?   The obvious exception is current `wait(float)`, but that is a long way from the normal user.

I already have `wait_ms(20)` to get 19-20ms.  I don't need `wait_us(20000)` to give me 19-20ms as well.

Personally, I'm really not a fan of size-of-argument-dependent behaviour here. Ideally something should be a sleeping thread-only function or a non-sleeping any-context function.

I would rather wait_us actually lost the sleep functionality altogether and became a non-sleeping function, and we maybe audited people using it with excessive values (warning/error on debug builds). I don't actually see the utility of `wait_us(20000)` with microsecond precision. 

I think we may be forced to have value-dependent behaviour on `wait(float)`, but I can live with that as long as there are alternative `wait_ms` and `wait_us` calls with solid behaviour.